src/plugins/data_explorer/public/components/sidebar/index.tsx

```diff
@@ -3,115 +3,95 @@
-  const viewOptions = useMemo(
-    () =>
-      views.map(({ id, title }) => ({
-        value: id,
-        text: title,
-      })),
-    [views]
-  );
+  const [selectedSources, setSelectedSources] = useState<DataSourceOption[]>([]);
+  const [dataSourceOptionList, setDataSourceOptionList] = useState<DataSourceGroup[]>([]);
+  const [activeDataSources, setActiveDataSources] = useState<DataSourceType[]>([]);
 
   const {
     services: {
-      data: { indexPatterns },
+      data: { indexPatterns, dataSources },
       notifications: { toasts },
+      application,
     },
   } = useOpenSearchDashboards<DataExplorerServices>();
 
   useEffect(() => {
     let isMounted = true;
-    const fetchIndexPatterns = async () => {
-      await indexPatterns.ensureDefaultIndexPattern();
-      const cache = await indexPatterns.getCache();
-      const currentOptions = (cache || []).map((indexPattern) => ({
-        label: indexPattern.attributes.title,
-        value: indexPattern.id,
-      }));
-      if (isMounted) {
-        setOptions(currentOptions);
+    const subscription = dataSources.dataSourceService.dataSources$.subscribe(
+      (currentDataSources) => {
+        if (isMounted) {
+          setActiveDataSources(Object.values(currentDataSources));
+        }
       }
-    };
-    fetchIndexPatterns();
+    );
 
     return () => {
+      subscription.unsubscribe();
       isMounted = false;
     };
-  }, [indexPatterns]);
+  }, [indexPatterns, dataSources]);
+
+  const getMatchedOption = (dataSourceList: DataSourceGroup[], ipId: string) => {
+    for (const dsGroup of dataSourceList) {
+      const matchedOption = dsGroup.options.find((item) => item.value === ipId);
+      if (matchedOption !== undefined) return matchedOption;
+    }
+    return undefined;
+  };
 
-  // Set option to the current index pattern
   useEffect(() => {
     if (indexPatternId) {
-      const option = options.find((o) => o.value === indexPatternId);
-      setSelectedOption(option);
+      const option = getMatchedOption(dataSourceOptionList, indexPatternId);
+      setSelectedSources(option ? [option] : []);
+    }
+  }, [indexPatternId, activeDataSources, dataSourceOptionList]);
+
+  const handleSourceSelection = (selectedDataSources: DataSourceOption[]) => {
+    if (selectedDataSources.length === 0) {
+      setSelectedSources(selectedDataSources);
+      return;
     }
-  }, [indexPatternId, options]);
+    // Temporary redirection solution for 2.11, where clicking non-index-pattern datasource
+    // will redirect user to Observability event explorer
+    if (selectedDataSources[0]?.ds?.getType() !== 'DEFAULT_INDEX_PATTERNS') {
+      return application.navigateToUrl(
+        `../observability-logs#/explorer?datasourceName=${selectedDataSources[0].label}&datasourceType=${selectedDataSources[0].type}`
+      );
+    }
+    setSelectedSources(selectedDataSources);
+    dispatch(setIndexPattern(selectedDataSources[0].value));
```

Comment:
We need guards for `selectedDataSources[0]`.

Reply:
there are already guards above
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/e8f219b3febaaa5d5d7ce21cd9c39dc560268e0e/src/plugins/data_explorer/public/components/sidebar/index.tsx#L60-L63

Reply:
I don't feel that that's a good-enough guard.   The happy path may be to always been "adding" to the array... but if the array gets reset to zero sometime in teh future then no update takes place.  There needs to be a guard on both sides.

Reply:
i don't think that's the right way to handle this, because `selectedDataSources` should be immutable. i think the root issue is `selectedDataSources` should be a tuple, not an array. i don't believe users can select multiple data sources?